### PR TITLE
structured_sources: fix subdir handling

### PR DESCRIPTION
**Pull request by dcbaker**
We currently don't handle subdirectories correctly in
structured_sources, which is problematic. To make this easier to handle
correctly, I've simply changed `structured_sources` to only use Files
and not strings as an implementation detail.

### CODE REVIEW

**Comment by tristan957** [comment]
This looks good to me. Will have to work through a merge conflict, but that is fine.

File: mesonbuild/backend/ninjabackend.py [hunk 1]
```diff
@@ -1721,9 +1719,13 @@ def generate_rust_target(self, target: build.BuildTarget) -> None:
...
                 else:
                     main_rust_file = os.path.join(g.get_subdir(), i)
-                orderdeps.extend([os.path.join(self.build_to_src, target.subdir, s)
-                                  for s in  target.structured_sources.as_list()])
 
+                for f in target.structured_sources.as_list():
+                    if isinstance(f, File):
+                        orderdeps.append(f.rel_to_builddir(self.build_to_src))
+                    else:
+                        orderdeps.extend([os.path.join(self.build_to_src, f.subdir, s)
```
[the remark this answers is not in the document]
Comment (eli-schwartz) [reply mesonbuild/backend/ninjabackend.py:1727]: GitHub won't let me review this context:
```
                if isinstance(g, str):
                    g = File.from_source_file(self.environment.source_dir, target.subdir, g)
```

Do we need this anymore? Can it ever be a str, now?

[the remark this answers is not in the document]
Comment (dcbaker) [reply mesonbuild/backend/ninjabackend.py:1727]: > Do we need this anymore? Can it ever be a str, now?

Yeah, I dropped that in the latest version.

> We used to only handle str here, but now we switched to handling both? But we also only store them in the first place as File, now. (Wait, what did we used to do if given a File?

Fail I think? I don't see how the old code could have worked. In one revision the `as_list` tried to stringify things, but I don't see how this worked after that :/

[the remark this answers is not in the document]
Comment (dcbaker) [reply mesonbuild/backend/ninjabackend.py:1727]: @eli-schwartz I've split this into two commits, and added tests for the fixups that I noticed.

File: mesonbuild/backend/ninjabackend.py [hunk 2]
```diff
@@ -1709,17 +1709,23 @@ def generate_rust_target(self, target: build.BuildTarget) -> None:
                 _ods, main_rust_file = self.__generate_compile_structure(target)
                 orderdeps.extend(_ods)
             else:
+                # The only way to get here is to have a only root files at are
```
Comment (eli-schwartz) [inline mesonbuild/backend/ninjabackend.py:1712]: Seems like there's a couple typos here. At least "at" -> "that", and "a only root files" is probably not supposed to have the leading "a"?

Comment (dcbaker) [reply mesonbuild/backend/ninjabackend.py]: I reworded it

File: mesonbuild/backend/ninjabackend.py [hunk 3]
```diff
@@ -1707,23 +1705,28 @@ def generate_rust_target(self, target: build.BuildTarget) -> None:
 
         main_rust_file = None
         if target.structured_sources:
-            if target.structured_sources.needs_copy(target):
+            if target.structured_sources.needs_copy():
                 _ods, main_rust_file = self.__generate_compile_structure(target)
                 orderdeps.extend(_ods)
             else:
+                # The only way to get here is to have only files in the "root"
```
Comment (tristan957) [inline mesonbuild/backend/ninjabackend.py:1712]: Should this comment maybe be a doc comment on the needs copy function?

**Review by eli-schwartz — APPROVED** [review]

**Pull request merged**